Sources/OpenCombine/Publishers/ChildManagingInnerBase.swift

```diff
@@ -0,0 +1,327 @@
+//
+//  ChildManagingInnerBase.swift
+//
+//  Created by Eric Patey on 21.08.2019.
+//
+
+/// This is an "Inner" base class that handles all of the management of child publishers
+/// and subscriptions for those publishers that make additional subscriptions beyond or
+///  instead of upstream.
+///
+/// Common subclasses are FlatMap and (soon to be Merge)
+internal class InnerWithChildrenBase<Child: Publisher, Downstream: Subscriber>
+    : Cancellable
+where Child.Output == Downstream.Input, Child.Failure == Downstream.Failure {
+    internal typealias Output = Downstream.Input
+    internal typealias Failure = Downstream.Failure
+    private typealias WorkType = (
+        value: Downstream.Input,
+        pausedChild: ChildSubscriber?)
+
+    // TODO: Implement CustomReflectable. See ref implementation in OperatorSubscription.
+
+    // Locking rules for this class.
+    //  - All mutable state must only be accessed while `lock` is held.
+    //  - In order to avoid any deadlock potential, it is absolutely forbidden to have
+    //      any sort of call out from this class while the lock is held. This is why
+    //      the draining of the work queue uses a relatively complex pattern.
+    private let lock = Lock(recursive: false)
+    private var downstream: Downstream?
+    private var childSubscribers = Set<ChildSubscriber>()
+    private var downstreamDemand = Subscribers.Demand.unlimited
+    private var valuesToSend = [WorkType]()
+    private var processingQueue = false
+    private var sendFinishedAfterDrainingQueue = false
+
+    internal init(downstream: Downstream) {
+        self.downstream = downstream
+    }
+
+    // These locked/unlocked functions are available to be overridden by subclasses
+    internal func lockedCancel() {}
+    internal func unlockedCancel() {}
+    internal func lockedDeactivate() {}
+    internal func unlockedDeactivate() {}
+    internal func lockedHasNonChildSubscription() -> Bool { return false }
+
+    internal final func sendSubscriptionDownstream(subscription: Subscription) {
+        downstream?.receive(subscription: subscription)
+    }
+
+    internal final func sendFailureDownstream(failure: Failure) {
+        downstream?.receive(completion: .failure(failure))
+        deactivate()
+    }
+
+    internal final func addChild(_ newChild: Child) {
+        let newChildSubscriber = ChildSubscriber(
+            receiveValue: { [weak self] in
```

Comment:
If you feel strongly, I'll make that change.

In general, though, I feel that it's just not worth that sort of analysis. The downsides of getting it wrong are huge, and the upside of using `unowned` is very minor. It just doesn't seem worth it to me...unless, we see data suggesting that the `weak` operations are material from a performance perspective.

Put a different way, whenever I see `unowned` while I'm doing a code review, it's a big flag and I have to spend a big amount of energy proving to myself that it's safe.

Reply:
It just creates dead code that will never be executed, which I find misleading.
I also believe that if invariants are broken, we should not propagate magic constants instead of telling that something went wrong.

If `ChildSubscriber` ever outlives `self` (maybe due to a bug that will be introduced later), it is our fault, not the client's, and we have better chances to spot it if there's a crash, and the client will not have to spend hours figuring out why something silently behaves wrong.